app/text_buffer.py

```diff
@@ -141,155 +143,222 @@ def doMoveLines(self, begin, end, to):
+        self.redo()
 
   def redoAddChange(self, change):
     """Push a change onto the end of the redoChain. Call redo() to enact the
         change."""
+    def noOpInstruction(change):
+      # Eliminate no-op entries
+      # app.log.info('opti', change)
+      noOpInstructions = set([
+        ('m', (0,0,0,0,0)),
+      ])
+      assert ('m', (0,0,0,0,0)) in noOpInstructions
+      if change in noOpInstructions:
+        return True
+      return False
+
+    newTrivialChange = False
     if self.debugRedo:
       app.log.info('redoAddChange', change)
     # When the redoChain is trimmed we may lose the saved at.
-    if self.redoIndex < self.savedAtRedoIndex:
-      self.savedAtRedoIndex = -1
-    self.redoChain = self.redoChain[:self.redoIndex]
+    # Trim only when there is a non-trivial action.
+    if change[0] == 'm' and self.redoIndex != len(self.redoChain):
+      newTrivialChange = True
+    else:
+      # Trim and combine main redoChain with tempChange
+      # if there is a non-trivial action
+      if self.redoIndex < self.savedAtRedoIndex:
+        self.savedAtRedoIndex = -1
+      self.redoChain = self.redoChain[:self.redoIndex]
+      if self.tempChange:
+        if self.redoChain[-1][0] == 'm':
+          combinedChange = (change[0], addVectors(self.tempChange[1], self.redoChain[-1][1]))
+          if noOpInstruction(combinedChange):
+            self.redoChain.pop()
+            self.redoIndex -= 1
+          else:
+            self.redoChain[self.redoIndex - 1] = combinedChange
+        else:
+          self.redoChain.append(self.tempChange)
+          self.redoIndex += 1
+        self.tempChange = None
     if 1: # optimizer
       if len(self.redoChain) and self.savedAtRedoIndex != self.redoIndex:
         if (self.redoChain[-1][0] == change[0] and
             change[0] in ('d', 'i')):
           change = (change[0], self.redoChain[-1][1] + change[1])
           self.undoOne()
           self.redoChain.pop()
-        elif change[0] == 'm':
-          if self.redoChain[-1][0] == 'm':
-            change = (change[0], addVectors(self.redoChain[-1][1], change[1]))
-            self.undoOne()
-            self.redoChain.pop()
         elif self.redoChain[-1][0] == change[0] and change[0] == 'n':
           change = (change[0], addVectors(self.redoChain[-1][1], change[1]))
           self.undoOne()
           self.redoChain.pop()
-    if 1:
-      # Eliminate no-op entries
-      noOpInstructions = set([
-        ('m', (0,0,0,0,0)),
-      ])
-      assert ('m', (0,0,0,0,0)) in noOpInstructions
-      if change in noOpInstructions:
+    if newTrivialChange:
+      if self.tempChange:
+        #Combine new change with the existing tempChange
+        change = (change[0], addVectors(self.tempChange[1], change[1]))
+        self.undoOne()
+      if noOpInstruction(change):
+        self.tempChange = None
         return
-      #app.log.info('opti', change)
-    self.redoChain.append(change)
+      self.redoDirty = True
```

Comment:
No. So the reason I have this flag is that when we perform a mouse action, the program is designed so that it calls redoAddChange and then a redo(). I edited redoAddChange so that it'll add the change to tempChange instead and then the next redo will make this change visible. However, if the user then performs a CTRL + Y and performs another redo, we want it to revert the changes caused by tempChange and perform a redo on the regular chain. However, at this time, tempChange still has a change assigned to it, so I needed the flag to determine whether the change is active or not.